### DISCUSSION TITLE
String format in non english cultures

### DISCUSSION THREAD

**Discussion by lenyturmel** [angularsen/UnitsNet#970] [Q&A]
Hi all

I have two questions about formatting units and quantities

1. I am not able to find how to format a quantity using the long name of the unit 
Example:
`Volume.FromLiters(1)
//I would prefer "1 liter" and not "1 L"`

2.  I am not able to find how add new cultures for the ToString of units
Example:
```
Thread.CurrentThread.CurrentUICulture = Language.Fr.ToCulture();
VolumeUnit.Liter.ToString();  //I would like to get "Litre" and not "Liter
```

Could you please help me and give some entry points or sample in order to do that ?
Thank you for your help (and thank you for this awesome library)

**Comment by angularsen** [comment]
Hi,

#### ToString with long unit name

There is no support for long form names yet, sorry. We could add support for it similar to how we define abbreviations for different cultures: https://github.com/angularsen/UnitsNet/blob/master/Common/UnitDefinitions/Volume.json#L8-L25. 

#### Localize unit name

Again, we support localization of abbreviations like "kg" and "cm", but currently no localization support for quantity name "Length" or unit name "Meter".


Would you be interested in attempting a pull request on either of these?
Relevant #890, #397 .

**Reply by lenyturmel** [reply]
Hi Andreas!
Thank you for your prompt answer.

Sorry, I had not seen the 2 tickets you mentionned.

Reading #890, #397, it seems there is a work in progress . Is there any branch for this work? Is there any sample code that illustrates how you would like to manage the localization of unit names and quantity type names?


I think I am going to develop a simple work around by my side and next I will try to propose you some code?

**Reply by lenyturmel** [reply]
Hi Andreas,

I've just took a look to the code to understand how you manage unit abbreviations. Reading the previous issues  #890, #397 and your suggestions,  I can see that the process to manage the localization of unit names should be very similar.

for the JSON data, I remain on :
` 

      "SingularName": "Meter",
      "PluralName": "Meters",
      "BaseUnits": {
        "L": "Meter"
      },
      "FromUnitToBaseFunc": "x",
      "FromBaseToUnitFunc": "x",
      "Prefixes": [ "Nano", "Micro", "Milli", "Centi", "Deci", "Hecto", "Kilo" ],
      "Localization": [
        {
          "Culture": "en-US",
          "Abbreviations": [ "m" ]
        },
        {
          "Culture": "ru-RU",
          "Abbreviations": [ "м" ]
        },
        {
          "Culture": "fr-FR",
          "Abbreviations": [ "м" ],
          **"SingularName": "Mètre",
          "PluralName": "Mètres"**
        }
`

I will try to add the expeccted properties in Localization to get the values from the JSON.
If you have any advice for helping to get started with  the implementation of the localization of unit names, I am more than interested :)

Thanks
Leny

**Reply by lenyturmel** [reply]
I have just started to write some code to manage the localization of unit names.

If you agree, could you please allow me to push my branch (I will push under the "XP" logical folder) in order to let you have a look to my work ?

I think I have to have similar classes  to UnitValueAbbreviationLookup and UnitAbbreviationsCache. Do you confirm ?  
I think we could rename UnitAbbreviationsCache as UnitNamesCache or UnitLocalizationCache and make this class to expose the following public methods:

- GetDefaultAbbreviation
- GetSingularName(TUnitType unit, IFormatProvider? formatProvider = null)
- GetPluralName(TUnitType unit, IFormatProvider? formatProvider = null)

For the lookup class, I  can see 2 options:

1. renaming UnitValueAbbreviationLookup  as UnitValueStringLookup (or any similar gebneric names) and use this class as un helper to store maps of string and units (string could be abbreviations, singular names or plural names). So, in UnitLocalizationCache, we will use 3 different UnitValueStringLookup : one for the abbreviation, one for the singular names and one for the plural names
2. renaming UnitValueAbbreviationLookup  as UnitValueLocalizartionLookup  and make this class to manage the three components : the abbreviations , the singular names and the plural names. So, in UnitLocalizationCache, we will use 1 UnitValueLocalizartionLookup  

What do you think about it?
Do you have any better idea?
Thank you for your help.

**Reply by angularsen** [reply]
@lenyturmel Please push a branch and create a pull request, so we can discuss details there.

> I think I have to have similar classes to UnitValueAbbreviationLookup and UnitAbbreviationsCache. Do you confirm ?

Yes, that sounds like a good approach.

> I think we could rename UnitAbbreviationsCache as UnitNamesCache or UnitLocalizationCache and make this class to expose the following public methods:

That would be fine, but we can't make breaking changes of public types and members without bumping major version, so either 
- Create a new class and mark the old one `[Obsolete("Use XXX instead.")]`
- Create a new class for this purpose alone and extract any private/internal code in the unit abbreviation classes if you want to reuse it.
- Live the wrong name for now, add a TODO comment and mention it in #180 for a future rename.

Without knowing any details, I would generally favor nr 2 or 3.

> For the lookup class, I can see 2 options:

It sounds like you are going in the right direction. Let's get a PR going and we have some code to look at and get a feel of. Also, having some unit tests to verify how it will be used and that it actually works can be very helpful. See how tests are written for the existing stuff. Don't need a ton of tests, just a few to verify your idea as a start.

**Reply by lenyturmel** [reply]
Hi Andreas
I tried to push my current branch with my work in progress but it looks like I do not have access rights to push branches :
`$ git push origin xp/localize-unit-names
Select an authentication method for 'https://github.com/':
  1. Web browser (default)
  2. Personal access token
option (enter for default): 1
info: please complete authentication in your browser...
remote: Permission to angularsen/UnitsNet.git denied to lenyturmel.
fatal: unable to access 'https://github.com/angularsen/UnitsNet.git/': The requested URL returned error: 403`

I tried with the Git console, and Githhub Deskstop but I have the same problem when I try to push my branch on your remote repo
![image](https://user-images.githubusercontent.com/53477637/135974980-2e16c2d8-79b9-4e29-8cc8-6a46a84d2cff.png)

I did never collaborate on a Github project (I work with AzureDevOps or Gitlab). You are not supposed to add me as a collaborator to give me such an authorization ?

Sorry for the troubles

**Reply by angularsen** [reply]
@lenyturmel No problem. You need to fork the repo and create a pull request from a branch on your fork to this repo.

Some guidance here:
https://github.com/angularsen/UnitsNet/wiki/Adding-a-New-Unit#6-create-pull-request

Please ask if you get stuck, I'm happy to assist.

**Reply by angularsen** [reply]
Direct link:
https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork

**Reply by lenyturmel** [reply]
Thank you for your help. I was able to fork the repo , to push my branch and to create a pull request as a draft (so you can follow my work)
I will try to work asap
thanks
